README.md
# clean-stacktrace [![NPM version](https://img.shields.io/npm/v/clean-stacktrace.svg?style=flat)](https://www.npmjs.com/package/clean-stacktrace) [![mit license][license-img]][license-url] [![NPM monthly downloads](https://img.shields.io/npm/dm/clean-stacktrace.svg?style=flat)](https://npmjs.org/package/clean-stacktrace) [![npm total downloads][downloads-img]][downloads-url]

> Clean up error stack traces from node internals

[![code climate][codeclimate-img]][codeclimate-url] 
[![code style][standard-img]][standard-url] 
[![linux build][travis-img]][travis-url] 
[![windows build][appveyor-img]][appveyor-url] 
[![code coverage][coverage-img]][coverage-url] 
[![dependency status][david-img]][david-url]
[![paypal donate][paypalme-img]][paypalme-url] 

You might also be interested in [stack-utils](https://github.com/tapjs/stack-utils#readme).

## Table of Contents
- [Install](#install)
- [Usage](#usage)
- [API](#api)
  * [cleanStacktrace](#cleanstacktrace)
- [Related](#related)
- [Contributing](#contributing)
- [Building docs](#building-docs)
- [Running tests](#running-tests)
- [Author](#author)
- [License](#license)

_(TOC generated by [verb](https://github.com/verbose/verb) using [markdown-toc](https://github.com/jonschlinkert/markdown-toc))_

## Install
Install with [npm](https://www.npmjs.com/)

```
$ npm install clean-stacktrace --save
```

or install using [yarn](https://yarnpkg.com)

```
$ yarn add clean-stacktrace
```

## Usage
> For more use-cases see the [tests](test.js)

```js
const cleanStacktrace = require('clean-stacktrace')
```

## API

### [cleanStacktrace](index.js#L58)
> Removes mostly not needed internal Nodejs entries. If you pass `mapper` function, you can make more changes to each line of the stack - for example making the paths to be relative, not absolute.

**Params**

* `stack` **{String}**: an error stack trace    
* `mapper` **{Function}**: more customization for each line    
* `returns` **{String}**: modified and cleaned stack  

**Example**

```js
var cleanStack = require('clean-stacktrace')
var error = new Error('Missing unicorn')

console.log(error.stack)
// =>
// Error: Missing unicorn
//     at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:2:15)
//     at Module._compile (module.js:409:26)
//     at Object.Module._extensions..js (module.js:416:10)
//     at Module.load (module.js:343:32)
//     at Function.Module._load (module.js:300:12)
//     at Function.Module.runMain (module.js:441:10)
//     at startup (node.js:139:18)

console.log(cleanStack(error.stack))
// =>
// Error: Missing unicorn
//     at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:2:15)

// or making paths relative
var path = require('path')
var stack = clean(error.stack, (line) => {
  var m = /.*\((.*)\).?/.exec(line) || []
  return m[1] ? line.replace(m[1], path.relative(process.cwd(), m[1])) : line
})
// =>
// Error: Missing unicorn
//     at Object.<anonymous> (unicorn.js:2:15)
```

## Related
- [always-done](https://www.npmjs.com/package/always-done): Handle completion and errors with elegance! Support for streams, callbacks, promises, child processes, async/await and sync functions. A drop-in replacement… [more](https://github.com/hybridables/always-done#readme) | [homepage](https://github.com/hybridables/always-done#readme "Handle completion and errors with elegance! Support for streams, callbacks, promises, child processes, async/await and sync functions. A drop-in replacement for [async-done][] - pass 100% of its tests plus more")
- [minibase](https://www.npmjs.com/package/minibase): Minimalist alternative for Base. Build complex APIs with small units called plugins. Works well with most of the already existing… [more](https://github.com/node-minibase/minibase#readme) | [homepage](https://github.com/node-minibase/minibase#readme "Minimalist alternative for Base. Build complex APIs with small units called plugins. Works well with most of the already existing [base][] plugins.")
- [try-catch-core](https://www.npmjs.com/package/try-catch-core): Low-level package to handle completion and errors of sync or asynchronous functions, using [once][] and [dezalgo][] libs. Useful for and… [more](https://github.com/hybridables/try-catch-core#readme) | [homepage](https://github.com/hybridables/try-catch-core#readme "Low-level package to handle completion and errors of sync or asynchronous functions, using [once][] and [dezalgo][] libs. Useful for and used in higher-level libs such as [always-done][] to handle completion of anything.")

## Contributing
Pull requests and stars are always welcome. For bugs and feature requests, [please create an issue](https://github.com/tunnckoCore/clean-stacktrace/issues/new).  
Please read the [contributing guidelines](CONTRIBUTING.md) for advice on opening issues, pull requests, and coding standards.  
If you need some help and can spent some cash, feel free to [contact me at CodeMentor.io](https://www.codementor.io/tunnckocore?utm_source=github&utm_medium=button&utm_term=tunnckocore&utm_campaign=github) too.

**In short:** If you want to contribute to that project, please follow these things

1. Please DO NOT edit [README.md](README.md), [CHANGELOG.md](CHANGELOG.md) and [.verb.md](.verb.md) files. See ["Building docs"](#building-docs) section.
2. Ensure anything is okey by installing the dependencies and run the tests. See ["Running tests"](#running-tests) section.
3. Always use `npm run commit` to commit changes instead of `git commit`, because it is interactive and user-friendly. It uses [commitizen][] behind the scenes, which follows Conventional Changelog idealogy.
4. Do NOT bump the version in package.json. For that we use `npm run release`, which is [standard-version][] and follows Conventional Changelog idealogy.

Thanks a lot! :)

## Building docs
Documentation and that readme is generated using [verb-generate-readme][], which is a [verb][] generator, so you need to install both of them and then run `verb` command like that

```
$ npm install verbose/verb#dev verb-generate-readme --global && verb
```

_Please don't edit the README directly. Any changes to the readme must be made in [.verb.md](.verb.md)._

## Running tests
Clone repository and run the following in that cloned directory

```
$ npm install && npm test
```

## Author
**Charlike Mike Reagent**

+ [github/tunnckoCore](https://github.com/tunnckoCore)
+ [twitter/tunnckoCore](https://twitter.com/tunnckoCore)
+ [codementor/tunnckoCore](https://codementor.io/tunnckoCore)

## License
Copyright © 2016-2017, [Charlike Mike Reagent](https://i.am.charlike.online). Released under the [MIT License](LICENSE).

***

_This file was generated by [verb-generate-readme](https://github.com/verbose/verb-generate-readme), v0.4.3, on March 10, 2017._  
_Project scaffolded using [charlike][] cli._

[always-done]: https://github.com/hybridables/always-done
[async-done]: https://github.com/gulpjs/async-done
[base]: https://github.com/node-base/base
[charlike]: https://github.com/tunnckocore/charlike
[commitizen]: https://github.com/commitizen/cz-cli
[dezalgo]: https://github.com/npm/dezalgo
[once]: https://github.com/isaacs/once
[standard-version]: https://github.com/conventional-changelog/standard-version
[verb-generate-readme]: https://github.com/verbose/verb-generate-readme
[verb]: https://github.com/verbose/verb

[license-url]: https://www.npmjs.com/package/clean-stacktrace
[license-img]: https://img.shields.io/npm/l/clean-stacktrace.svg

[downloads-url]: https://www.npmjs.com/package/clean-stacktrace
[downloads-img]: https://img.shields.io/npm/dt/clean-stacktrace.svg

[codeclimate-url]: https://codeclimate.com/github/tunnckoCore/clean-stacktrace
[codeclimate-img]: https://img.shields.io/codeclimate/github/tunnckoCore/clean-stacktrace.svg

[travis-url]: https://travis-ci.org/tunnckoCore/clean-stacktrace
[travis-img]: https://img.shields.io/travis/tunnckoCore/clean-stacktrace/master.svg?label=linux

[appveyor-url]: https://ci.appveyor.com/project/tunnckoCore/clean-stacktrace
[appveyor-img]: https://img.shields.io/appveyor/ci/tunnckoCore/clean-stacktrace/master.svg?label=windows

[coverage-url]: https://codecov.io/gh/tunnckoCore/clean-stacktrace
[coverage-img]: https://img.shields.io/codecov/c/github/tunnckoCore/clean-stacktrace/master.svg

[david-url]: https://david-dm.org/tunnckoCore/clean-stacktrace
[david-img]: https://img.shields.io/david/tunnckoCore/clean-stacktrace.svg

[standard-url]: https://github.com/feross/standard
[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg

[paypalme-url]: https://www.paypal.me/tunnckoCore
[paypalme-img]: https://img.shields.io/badge/paypal-donate-brightgreen.svg



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.1.0"></a>
# [1.1.0](https://github.com/tunnckocore/clean-stacktrace/compare/v1.0.0...v1.1.0) (2017-03-10)


### Bug Fixes

* **changelog:** update, because was overwritten ([07cb52c](https://github.com/tunnckocore/clean-stacktrace/commit/07cb52c))


### Features

* **refactor:** rewrite to use stack-utils-node-internals ([e669e42](https://github.com/tunnckocore/clean-stacktrace/commit/e669e42))





## 1.0.0 - 2016-11-04
- First stable release
- semantic versioning
- port to work on node.js 0.10

## 0.0.0 - 2016-11-04
- Initial commit


test.js
/*!
 * clean-stacktrace <https://github.com/tunnckoCore/clean-stacktrace>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (https://i.am.charlike.online)
 * Released under the MIT license.
 */

/* jshint asi:true */

'use strict'

var path = require('path')
var test = require('mukla')
var clean = require('./index')

test('default', function (done) {
  var pre = [
    'Error: foo',
    '    at Test.fn (/Users/sindresorhus/dev/clean-stack/test.js:6:15)'
  ]
  var stack = pre.concat([
    '    at handleMessage (internal/child_process.js:695:10)',
    '    at Pipe.channel.onread (internal/child_process.js:440:11)',
    '    at process.emit (events.js:172:7)'
  ])

  test.strictEqual(clean(stack), pre.join('\n'))
  done()
})

test('default #2', function (done) {
  var pre = [
    'Error: foo',
    '    at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:4:7)'
  ]
  var stack = pre.concat([
    '    at Module._compile (module.js:409:26)',
    '    at Object.Module._extensions..js (module.js:416:10)',
    '    at Module.load (module.js:343:32)',
    '    at Function.Module._load (module.js:300:12)',
    '    at Function.Module.runMain (module.js:441:10)',
    '    at startup (node.js:139:18)'
  ]).join('\n')
  test.ok(clean(stack) === pre.join('\n'), 'should be equal')
  done()
})

test('directly executed node script', function (done) {
  var pre = [
    'Error: foo',
    '    at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:4:7)'
  ]
  var stack = pre.concat([
    '    at Module._compile (module.js:409:26)',
    '    at Object.Module._extensions..js (module.js:416:10)',
    '    at Module.load (module.js:343:32)',
    '    at Function.Module._load (module.js:300:12)',
    '    at Function.Module.runMain (module.js:441:10)',
    '    at startup (node.js:139:18)',
    '    at node.js:968:3'
  ]).join('\n')
  var actual = clean(stack)
  var expected = pre.join('\n')
  test.strictEqual(actual, expected)
  done()
})

test('internal child_process', function (done) {
  var pre = [
    'Error: foo',
    '    at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:4:7)'
  ]
  test.strictEqual(clean(pre.concat([
    '    at Module._compile (module.js:409:26)',
    '    at Object.Module._extensions..js (module.js:416:10)',
    '    at internal/child_process.js:696:12'
  ])), pre.join('\n'))
  done()
})

test('babel-polyfill', function (done) {
  var pre = '    at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:4:7)'
  var stack = [
    pre,
    '    at run (/Users/sindresorhus/dev/clean-stack/node_modules/babel-polyfill/node_modules/core-js/modules/es6.promise.js:87:22)',
    '    at /Users/sindresorhus/dev/clean-stack/node_modules/babel-polyfill/node_modules/core-js/modules/es6.promise.js:100:28'
  ].join('\n')
  test.strictEqual(clean(stack), pre)
  done()
})

test('work on Windows', function (done) {
  var expected = 'Error: foo\n    at Test.fn (/Users/sindresorhus/dev/clean-stack/test.js:6:15)'
  var parts = [
    'Error: foo',
    '    at Test.fn (\\Users\\sindresorhus\\dev\\clean-stack\\test.js:6:15)',
    '    at handleMessage (internal\\child_process.js:695:10)',
    '    at Pipe.channel.onread (internal\\child_process.js:440:11)',
    '    at process.emit (events.js:172:7)'
  ]
  test.ok(clean(parts.join('\n')), expected)
  done()
})

test('should allow to use a mapper function', function (done) {
  var error = new Error('fixture err')
  var stack = clean(error.stack, (line) => {
    var m = /.*\((.*)\).*/.exec(line) || []
    return m[1] ? line.replace(m[1], path.relative(process.cwd(), m[1])) : line
  })

  test.ok(/\(test\.js:/.test(stack))
  done()
})


index.js
/*!
 * clean-stacktrace <https://github.com/tunnckoCore/clean-stacktrace>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (https://i.am.charlike.online)
 * Released under the MIT license.
 */

'use strict'

var regex = /(?:(?:(?:node|(?:internal\/|.*node_modules\/babel-polyfill\/.*)\w+)\.js:\d+:\d+)|native)/
var nodeInternals = require('stack-utils-node-internals')

/**
 * > Removes mostly not needed internal Nodejs entries.
 * If you pass `mapper` function, you can make more
 * changes to each line of the stack - for example making
 * the paths to be relative, not absolute.
 *
 * **Example**
 *
 * ```js
 * var cleanStack = require('clean-stacktrace')
 * var error = new Error('Missing unicorn')
 *
 * console.log(error.stack)
 * // =>
 * // Error: Missing unicorn
 * //     at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:2:15)
 * //     at Module._compile (module.js:409:26)
 * //     at Object.Module._extensions..js (module.js:416:10)
 * //     at Module.load (module.js:343:32)
 * //     at Function.Module._load (module.js:300:12)
 * //     at Function.Module.runMain (module.js:441:10)
 * //     at startup (node.js:139:18)
 *
 * console.log(cleanStack(error.stack))
 * // =>
 * // Error: Missing unicorn
 * //     at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:2:15)
 *
 * // or making paths relative
 * var path = require('path')
 * var stack = clean(error.stack, (line) => {
 *   var m = /.*\((.*)\).?/.exec(line) || []
 *   return m[1] ? line.replace(m[1], path.relative(process.cwd(), m[1])) : line
 * })
 * // =>
 * // Error: Missing unicorn
 * //     at Object.<anonymous> (unicorn.js:2:15)
 * ```
 *
 * @param  {String} `stack` an error stack trace
 * @param  {Function} `mapper` more customization for each line
 * @return {String} modified and cleaned stack
 * @api public
 */

module.exports = function cleanStacktrace (stack, mapper) {
  if (!Array.isArray(stack)) {
    stack = stack.split('\n')
  }

  var result = []
  var internals = nodeInternals().concat(regex)

  if (!(/^\s*at /.test(stack[0])) && (/^\s*at /i.test(stack[1]))) {
    result.push(stack[0])
    stack = stack.slice(1)
  }

  stack.forEach(function (line) {
    var isInternal = internals.some(function (internal) {
      return internal.test(line)
    })

    if (isInternal) {
      return null
    }

    result.push(line)
  })

  result = typeof mapper === 'function' ? result.map(mapper) : result

  return result.join('\n')
}
